Use CMake for build, auto-generate generated files, update to Vulkan-Headers 1.2.182

rbsheth opened the pull request:
I was looking at incorporating [`libktx`](https://github.com/KhronosGroup/KTX-Software/tree/master/lib) into various packages using the [Hunter](https://github.com/cpp-pm/hunter) package manager. I noticed that this repository was used as a subrepo in `libktx`, but had a lot of build system and file splitting across both repos that made it hard to compartmentalize.

Changes:
1) Consolidate the `Makefile` and `dfdutils.gyp` functionality into a new CMake-based build that supports Linux, Windows, and Mac. 
2) Added Travis CI testing for Windows.
3) Fetch KhronosGroup/DataFormat header and Vulkan-Headers from URLs and then patch them using CMake. This makes it easier to update and less confusing as to which versions are being used when integrated into other projects.
4) Add `install` and `export` functionality that allows for the use of `dfdutils` as an external project/library in CMake.

I plan to use this library inside a custom build for `libktx`. If there is interest, I can create a PR there for CMake improvements.

PTAL @MarkCallow

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/dfdutils?pullRequest=4) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/dfdutils?pullRequest=4) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/dfdutils?pullRequest=4) it.</sub>

rbsheth commented:
We should squash merge this so it isn't so messy. Otherwise, I can clean up the commit history.

MarkCallow commented:
Thank you for this. Providing a CMake-based build for Linux, Windows and Mac for this repo was on my todo list. I'll need to take a careful look which I won't have time to do for several days. Sorry.

What do you mean by "had a lot of build system and file splitting across both repos that made it hard to compartmentalize?" What about the other projects that are included in libktx via `git subrepo`? Do you have similar problems with them?

I'd like to know more about how you will use this library in libktx to help me decide whether to ask  you to create a PR there.

A few specific points based on your comments:

1. I'll squash merge this when the time comes, so don't worry about cleaning it up.
2. We deliberately have a copy of vulkan_core.h in this repo. It should not be replaced by the released header file. It has VkFormat enums for ASTC 3D texture formats. These enums have been assigned by the Vulkan WG but are not included in the released `vulkan_core.h` because as yet there is no implementation of the ASTC 3D texture extension. Fingers crossed there will be one soon. We need these enums in dfdutils and libktx because there are implementations of the OpenGL {,ES} extensions for ASTC 3D.
3. We want to move away from TravisCI. Would prefer GitHub Actions or Appveyor.

rbsheth commented:
Hello,

> Thank you for this. Providing a CMake-based build for Linux, Windows and Mac for this repo was on my todo list. I'll need to take a careful look which I won't have time to do for several days. Sorry.

No problem, happy if it gets in within the next couple weeks. 😄 
 
> What do you mean by "had a lot of build system and file splitting across both repos that made it hard to compartmentalize?" What about the other projects that are included in libktx via `git subrepo`? Do you have similar problems with them?
>
> I'd like to know more about how you will use this library in libktx to help me decide whether to ask you to create a PR there.

Yes, typically any third party code that could be shared with other libraries needs to be factored out into its own repo/library to work properly in Hunter. Just like how `apt` or `brew` works, Hunter is supposed to be a C++ package manager.

For example, I forked https://github.com/cpp-pm/basis_universal a while ago and refactored its CMakeLists to find each library ignoring the checked in sources (cf. https://github.com/cpp-pm/basis_universal/blob/hunter-1.15/cmake/Config.cmake.in#L3). From what I can see, libktx depends on basis_universal, dfdutils (when Vulkan is enabled), and OpenGL-Registry (when OpenGL is enabled). Hunter will pull the transitive dependencies for basis_universal automatically (miniz, zstd, jpeg-compressor, etc.).

You can see a WIP branch here: https://github.com/cpp-pm/KTX-Software/blob/hunter-4.0.0-efc9f09/CMakeLists.txt#L116 and I deleted most third-party code I could find in this commit: https://github.com/cpp-pm/KTX-Software/commit/24e4e8a7963273c2e59e3983e04ac5253e39ce13

In dfdutils in particular, the `mkvk.cmake` script was in the KTX-Software repo but actually operates only on files relevant to dfdutils. So I moved it here and ported it to work correctly.

> 
> A few specific points based on your comments:
> 
>     1. I'll squash merge this when the time comes, so don't worry about cleaning it up.
👍 
> 
>     2. We deliberately have a copy of vulkan_core.h in this repo. It should not be replaced by the released header file. It has VkFormat enums for ASTC 3D texture formats. These enums have been assigned by the Vulkan WG but are not included in the released `vulkan_core.h` because as yet there is no implementation of the ASTC 3D texture extension. Fingers crossed there will be one soon. We need these enums in dfdutils and libktx because there are implementations of the OpenGL {,ES} extensions for ASTC 3D.

Yes, I noticed that. While we wait for the update, I converted your manual patch process into an automatic one using `.patch` files. This is the patch that applies to Vulkan-Headers v1.2.182 (https://github.com/rbsheth/dfdutils/blob/hunter-0.0.0-659a739/third_party/vulkan_core.patch). CMake downloads that release from Github (when not using Hunter) directly here: https://github.com/rbsheth/dfdutils/blob/hunter-0.0.0-659a739/CMakeLists.txt#L39 and the patch is actually applied here: https://github.com/rbsheth/dfdutils/blob/hunter-0.0.0-659a739/cmake/mkvk.cmake#L65. The original Vulkan SDK/Headers are NOT exposed to the build process, only the modified ones. Same with DataFormat.

It seems `vkformat_enum.h` is enough to propagate those enums to libktx, so I install that along side `dfd.h` and build the library. This works well and allows me to build `libktx` without any of the generated/patched files checked in. 

> 
>     3. We want to move away from TravisCI. Would prefer GitHub Actions or Appveyor.
> 
Yes, we did this with Hunter last year. Was a bit annoying at first but not too bad. I can help move CI after this PR is done.
> 
> I'm happy you are adding this and libktx to Hunter.
Thanks! We are happy to be using it and with the work you are doing.

MarkCallow commented:
> For example, I forked https://github.com/cpp-pm/basis_universal a while ago and refactored its CMakeLists to find each library ignoring the checked in sources (cf. https://github.com/cpp-pm/basis_universal/blob/hunter-1.15/cmake/Config.cmake.in#L3). From what I can see, libktx depends on basis_universal, dfdutils (when Vulkan is enabled), and OpenGL-Registry (when OpenGL is enabled). Hunter will pull the transitive dependencies for basis_universal automatically (miniz, zstd, jpeg-compressor, etc.).

dfdutils is needed for KTX v2, a need unrelated to Vulkan. It seems related because KTX v2 uses Vulkan format enums.

We adding a new dependency on astc-encoder. I suppose you'll want to make a lib of this too. I expect to add at least 2 more encoders in future.

I don't know if any package managers support these encoders. That is one of the reasons for including things as subrepos. Another reason is that some of them, e.g. basis_universal, do not build libraries.

> 
> You can see a WIP branch here: https://github.com/cpp-pm/KTX-Software/blob/hunter-4.0.0-efc9f09/CMakeLists.txt#L116 and I deleted most third-party code I could find in this commit: [cpp-pm/KTX-Software@24e4e8a](https://github.com/cpp-pm/KTX-Software/commit/24e4e8a7963273c2e59e3983e04ac5253e39ce13)
> 
> In dfdutils in particular, the `mkvk.cmake` script was in the KTX-Software repo but actually operates only on files relevant to dfdutils. So I moved it here and ported it to work correctly.

Be careful here. The file vkformat_enum.h which is used in these is actually KTX v2 specific. The mkvkformatfiles which creates that (and vkformat_str.c) from vulkan_core.h removes formats that are listed in the spec. as invalid in KTX v2.

`dfd2vk.inl` and `vk2dfd.inl` in dfdutils should include all Vulkan formats and KTX-Software should have custom versions. The header file to use is passed as a parameter to the script to make it easy to created different versions.

> 
> > A few specific points based on your comments:
> 
> 👍
> 
> > ```
> > 2. We deliberately have a copy of vulkan_core.h in this repo. It should not be replaced by the released header file. It has VkFormat enums for ASTC 3D texture formats. These enums have been assigned by the Vulkan WG but are not included in the released `vulkan_core.h` because as yet there is no implementation of the ASTC 3D texture extension. Fingers crossed there will be one soon. We need these enums in dfdutils and libktx because there are implementations of the OpenGL {,ES} extensions for ASTC 3D.
> > ```
> 
> Yes, I noticed that. While we wait for the update, I converted your manual patch process into an automatic one using `.patch` files. This is the patch that applies to Vulkan-Headers v1.2.182 (https://github.com/rbsheth/dfdutils/blob/hunter-0.0.0-659a739/third_party/vulkan_core.patch). CMake downloads that release from Github (when not using Hunter) directly here: https://github.com/rbsheth/dfdutils/blob/hunter-0.0.0-659a739/CMakeLists.txt#L39 and the patch is actually applied here: https://github.com/rbsheth/dfdutils/blob/hunter-0.0.0-659a739/cmake/mkvk.cmake#L65. The original Vulkan SDK/Headers are NOT exposed to the build process, only the modified ones. Same with DataFormat.

Actually the original patch process wasn't manual. The vulkan_core.h was generated using the normal Vulkan registry tools on an internal branch with the putative ASTC 3D extension enabled.

> 
> It seems `vkformat_enum.h` is enough to propagate those enums to libktx, so I install that along side `dfd.h` and build the library. This works well and allows me to build `libktx` without any of the generated/patched files checked in.

vkformat_enums is wrong for dfd, as explained above. The script that generates that also generates `vkformat_str.c` which is needed in KTX-Software.

rbsheth commented:
> > For example, I forked https://github.com/cpp-pm/basis_universal a while ago and refactored its CMakeLists to find each library ignoring the checked in sources (cf. https://github.com/cpp-pm/basis_universal/blob/hunter-1.15/cmake/Config.cmake.in#L3). From what I can see, libktx depends on basis_universal, dfdutils (when Vulkan is enabled), and OpenGL-Registry (when OpenGL is enabled). Hunter will pull the transitive dependencies for basis_universal automatically (miniz, zstd, jpeg-compressor, etc.).
> 
> dfdutils is needed for KTX v2, a need unrelated to Vulkan. It seems related because KTX v2 uses Vulkan format enums.

Okay, I can make sure to pull it in regardless of options set.
> 
> We adding a new dependency on astc-encoder. I suppose you'll want to make a lib of this too. I expect to add at least 2 more encoders in future.

Yes, we already have astc-encoder https://github.com/cpp-pm/astc-encoder. Which other encoders are you thinking about?
> 
> I don't know if any package managers support these encoders. That is one of the reasons for including things as subrepos. Another reason is that some of them, e.g. basis_universal, do not build libraries.
> 

Yes, I am not forcing users into a specific package manager. However, the option should be available to use one and consume the repo as a CMake target.

> > You can see a WIP branch here: https://github.com/cpp-pm/KTX-Software/blob/hunter-4.0.0-efc9f09/CMakeLists.txt#L116 and I deleted most third-party code I could find in this commit: [cpp-pm/KTX-Software@24e4e8a](https://github.com/cpp-pm/KTX-Software/commit/24e4e8a7963273c2e59e3983e04ac5253e39ce13)
> > In dfdutils in particular, the `mkvk.cmake` script was in the KTX-Software repo but actually operates only on files relevant to dfdutils. So I moved it here and ported it to work correctly.
> 
> Be careful here. The file vkformat_enum.h which is used in these is actually KTX v2 specific. The mkvkformatfiles which creates that (and vkformat_str.c) from vulkan_core.h removes formats that are listed in the spec. as invalid in KTX v2.
> 
> `dfd2vk.inl` and `vk2dfd.inl` in dfdutils should include all Vulkan formats and KTX-Software should have custom versions. The header file to use is passed as a parameter to the script to make it easy to created different versions.

Maybe the solution here is a CMake argument? Something like `DFDUTILS_BUILD_FOR_KTX` that runs the patching process only for building KTX software? It seemed like dfdutils won't build without the patched files.

> 
> > > A few specific points based on your comments:
> > 
> > 
> > 👍
> > > ```
> > > 2. We deliberately have a copy of vulkan_core.h in this repo. It should not be replaced by the released header file. It has VkFormat enums for ASTC 3D texture formats. These enums have been assigned by the Vulkan WG but are not included in the released `vulkan_core.h` because as yet there is no implementation of the ASTC 3D texture extension. Fingers crossed there will be one soon. We need these enums in dfdutils and libktx because there are implementations of the OpenGL {,ES} extensions for ASTC 3D.
> > > ```
> > 
> > 
> > Yes, I noticed that. While we wait for the update, I converted your manual patch process into an automatic one using `.patch` files. This is the patch that applies to Vulkan-Headers v1.2.182 (https://github.com/rbsheth/dfdutils/blob/hunter-0.0.0-659a739/third_party/vulkan_core.patch). CMake downloads that release from Github (when not using Hunter) directly here: https://github.com/rbsheth/dfdutils/blob/hunter-0.0.0-659a739/CMakeLists.txt#L39 and the patch is actually applied here: https://github.com/rbsheth/dfdutils/blob/hunter-0.0.0-659a739/cmake/mkvk.cmake#L65. The original Vulkan SDK/Headers are NOT exposed to the build process, only the modified ones. Same with DataFormat.
> 
> Actually the original patch process wasn't manual. The vulkan_core.h was generated using the normal Vulkan registry tools on an internal branch with the putative ASTC 3D extension enabled.

I see. Any harm in doing it this way with a patch file? I diffed the output and found it to be the same (minus an extension definition at the bottom of the header) for the 152 headers.

> 
> > It seems `vkformat_enum.h` is enough to propagate those enums to libktx, so I install that along side `dfd.h` and build the library. This works well and allows me to build `libktx` without any of the generated/patched files checked in.
> 
> vkformat_enums is wrong for dfd, as explained above. The script that generates that also generates `vkformat_str.c` which is needed in KTX-Software.

Same comment as above, maybe a CMake argument?

MarkCallow commented:
My review and responses have been delayed by the mudslide in Atami cutting off all lifeline services to our house. Still waiting for water & internet.

One quick comment: the idea of the cmake variable to build for KTX seems good. I need to refresh my memory about the formats forbidden in KTX2. Some of them are because a DFD can’t distinguish between them and others. Those must not be included in dfdutils either.

MarkCallow commented on CMakeLists.txt:
```diff
@@ -0,0 +1,191 @@
+    vk2dfd.c
+    ${GENERATED_DIR}/vk2dfd.inl
+    ${GENERATED_DIR}/vkformat_check.c
+    ${GENERATED_DIR}/vkformat_str.c
```
These 2, especially vkformat_check.c, are not truly part of the dfdutils. As I mentioned elsewhere I need to refresh my memory re. the forbidden formats to see if a similar file is needed for dfdutils. I think not because vk2dfd.inl will reflect any unsupported formats. KTX-Software uses the file to check vkformat values in incoming .KTX2 files.

MarkCallow replied:
I've checked the KTX v2 prohibited formats. There are 3 classes of them. The first 2 classes should all be handled by vk2dfd. The 3rd class is multisample formats which are not supported by KTX but can be described by a DFD. I think vk2dfd probably doesn't handle them at present but it could. vkformat_check.c prohibits all 3 classes so it does not belong in dfdutils. Leave it to KTX-Software.

I am debating with myself about vkformat_str. An error message using a format string would be emitted by a caller of dfdutils not by anything in dfdutils which suggests removing it too. But having it here removes the need for the each caller to create something similar so I can see reason to leave it here. But they are both generated by the same awk script so I think it best to remove vkformat_str too.

MarkCallow commented:
I've let this slide too long, in part due to our local disaster.

Please fix the build issue whatever it is, and remove the `vkformat*` files and the script that generates them.

MarkCallow commented:
Please fix this by 2022.1.31 as previously requested, otherwise I will close this PR.

MarkCallow commented:
Regrettably must close this due to inaction. @rbsheth if your interest returns and you want to fix the PR, please reopen it.
